### PR TITLE
Enable clip broadcasting, with apply_ufunc

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/5173
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] Add docstring
- [x] Confirm apply_ufunc args, like join

I think this is the right approach — we may need to confirm the appropriate args to `apply_ufunc` — probably we want the same as `.where`?

Tests are OK but a bit lacking atm.

### CODE REVIEW

File: xarray/core/common.py [hunk 1]
```diff
@@ -380,6 +380,16 @@ def squeeze(
...
 
+    def clip(self, min=None, max=None, *, keep_attrs=None):
+        from .computation import apply_ufunc
+
+        if keep_attrs is None:
+            # When this was a unary func, the default was True, so retaining the
+            # default.
+            keep_attrs = _get_keep_attrs(default=True)
+
+        return apply_ufunc(np.clip, self, min, max, keep_attrs=keep_attrs)
```
Comment (dcherian) [inline xarray/core/common.py:391]: LGTM. We could also solve #4796 by setting `dask="allowed"` here.

File: xarray/tests/test_dataarray.py [hunk 2]
```diff
@@ -7313,3 +7313,18 @@ def test_deepcopy_obj_array():
...
+
+    result = da.clip(max=0.5)
+    assert result.max(...) <= 0.5
+
+    result = da.clip(min=0.25, max=0.75)
+    assert result.min(...) >= 0.25
+    assert result.max(...) <= 0.75
+
+    result = da.clip(min=da.mean("x"), max=da.mean("a"))
+    assert result.dims == da.dims
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:7330]: How about 

```
assert_array_equal(result.data, np.clip(da.data, min=da.mean("x").data, max=da.mean("a").data))
```

Comment (max-sixty) [reply xarray/tests/test_dataarray.py:7330]: Perfect

File: xarray/core/variable.py [hunk 3]
```diff
@@ -1665,6 +1665,11 @@ def fillna(self, value):
     def where(self, cond, other=dtypes.NA):
         return ops.where_method(self, cond, other)
 
+    def clip(self, min=None, max=None):
+        from .computation import apply_ufunc
+
+        return apply_ufunc(np.clip, self, min, max)
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/variable.py:1671]: Thanks, added (though maybe they need to be filled out a bit)

**Comment by max-sixty** [comment]
The tests are in a reasonable state — not perfect — but maybe good enough to proceed. Are there any cases that are missing currently?

File: xarray/tests/test_dataarray.py [hunk 4]
```diff
@@ -7313,3 +7326,31 @@ def test_deepcopy_obj_array():
     x0 = DataArray(np.array([object()]))
     x1 = deepcopy(x0)
     assert x0.values[0] is not x1.values[0]
+
+
+def test_clip(da):
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:7331]:
```suggestion
@pytest.mark.parametrize("backend", ["numpy", "dask"])
def test_clip(da):
```

to test dask?

File: xarray/tests/test_dataarray.py [hunk 5]
```diff
@@ -7313,3 +7326,31 @@ def test_deepcopy_obj_array():
...
+    result = da.clip(min=0.5)
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:7332]:
```suggestion
    with raise_if_dask_computes():
    	result = da.clip(min=0.5)
```

and so on?

**Review by dcherian — APPROVED** [review]
Looks good to me though dask tests are missing. We can add them in a future PR if you don't want to do it now.

File: xarray/tests/test_dataarray.py (hunk 4 above)
Comment (max-sixty) [reply xarray/tests/test_dataarray.py:7331]: This already happens though?

Comment (dcherian) [reply xarray/tests/test_dataarray.py:7331]: Oh..... I see. You've explicitly added it only for `rolling_exp` which can only accept numpy. Got it

**Comment by max-sixty** [comment]
> Looks good to me though dask tests are missing. We can add them in a future PR if you don't want to do it now.

They're there! The magic of fixtures is `da` is attempted with both a dask and numpy backend: https://github.com/pydata/xarray/pull/5184/files#diff-ca5c2de2fe6e9e25fbf22bd53e4976c15da74900dfb14deb7e6e87f5377230e3R6438

Does that make sense, assuming I'm not missing anything? Too magical?

**Pull request merged**